File: app/AdbClient.cpp
#include "AdbClient.h"

#include <QFileInfo>
#include <QProcess>
#include <QRegularExpression>
#include <stdexcept>

namespace {
QString findDefaultAdb()
{
    const QStringList candidates = {
        QStringLiteral("C:/Program Files/BlueStacks_nxt/HD-Adb.exe"),
        QStringLiteral("C:/Program Files/BlueStacks/HD-Adb.exe"),
    };
    for (const QString &c : candidates) {
        if (QFileInfo::exists(c))
            return c;
    }
    return QStringLiteral("adb");
}
} // namespace

AdbClient::AdbClient(QString adbPath)
    : m_adbPath(adbPath.isEmpty() ? findDefaultAdb() : std::move(adbPath))
{
}

void AdbClient::setSerial(const QString &serial)
{
    m_serial = serial;
}

QByteArray AdbClient::run(const QStringList &args, int timeoutMs, bool check) const
{
    QStringList full = args;
    if (!m_serial.isEmpty())
        full.prepend(m_serial), full.prepend(QStringLiteral("-s"));

    QProcess p;
    p.start(m_adbPath, full);
    if (!p.waitForStarted(5000))
        throw std::runtime_error("cannot start adb: " + m_adbPath.toStdString());
    if (!p.waitForFinished(timeoutMs)) {
        p.kill();
        throw std::runtime_error("adb timeout");
    }
    if (check && p.exitCode() != 0) {
        const QString err = QString::fromUtf8(p.readAllStandardError());
        throw std::runtime_error(("adb failed: " + err).toStdString());
    }
    return p.readAllStandardOutput();
}

QStringList AdbClient::devices() const
{
    // Temporarily ignore serial for devices listing
    AdbClient tmp(m_adbPath);
    const QByteArray out = tmp.run({QStringLiteral("devices")});
    QStringList result;
    const QString text = QString::fromUtf8(out);
    const QStringList lines = text.split(QRegularExpression(QStringLiteral("[\r\n]+")), Qt::SkipEmptyParts);
    for (int i = 1; i < lines.size(); ++i) {
        const QStringList parts = lines[i].split(QRegularExpression(QStringLiteral("\\s+")), Qt::SkipEmptyParts);
        if (parts.size() >= 2 && parts[1] == QLatin1String("device"))
            result.push_back(parts[0]);
    }
    return result;
}

QString AdbClient::ensureDevice()
{
    if (!m_serial.isEmpty())
        return m_serial;
    const QStringList devs = devices();
    if (devs.isEmpty())
        throw std::runtime_error("No ADB device online (start BlueStacks)");
    for (const QString &d : devs) {
        if (d.startsWith(QLatin1String("emulator-"))) {
            m_serial = d;
            return m_serial;
        }
    }
    m_serial = devs.first();
    return m_serial;
}

QByteArray AdbClient::screencapPng()
{
    ensureDevice();
    QByteArray data = run({QStringLiteral("exec-out"), QStringLiteral("screencap"), QStringLiteral("-p")}, 60000);
    if (data.size() < 100)
        throw std::runtime_error("screencap returned empty data");
    static const QByteArray kPng = QByteArray::fromHex("89504e470d0a1a0a");
    if (!data.startsWith(kPng)) {
        data.replace("\r\n", "\n");
        if (!data.startsWith(kPng))
            throw std::runtime_error("screencap did not return a PNG");
    }
    return data;
}

void AdbClient::tap(int x, int y)
{
    ensureDevice();
    run({QStringLiteral("shell"), QStringLiteral("input"), QStringLiteral("tap"),
         QString::number(x), QString::number(y)});
}

void AdbClient::swipe(int x1, int y1, int x2, int y2, int durationMs)
{
    ensureDevice();
    run({QStringLiteral("shell"), QStringLiteral("input"), QStringLiteral("swipe"),
         QString::number(x1), QString::number(y1), QString::number(x2), QString::number(y2),
         QString::number(durationMs)});
}

std::pair<int, int> AdbClient::wmSize()
{
    ensureDevice();
    const QString text = QString::fromUtf8(run({QStringLiteral("shell"), QStringLiteral("wm"), QStringLiteral("size")}));
    const QRegularExpression re(QStringLiteral("(\\d+)x(\\d+)"));
    const auto m = re.match(text);
    if (!m.hasMatch())
        throw std::runtime_error("cannot parse wm size");
    return {m.captured(1).toInt(), m.captured(2).toInt()};
}

File: src/autoarchero/adb_client.py
from __future__ import annotations

import shutil
import subprocess
from pathlib import Path
from typing import List, Optional


DEFAULT_ADB_CANDIDATES = [
    Path(r"C:\Program Files\BlueStacks_nxt\HD-Adb.exe"),
    Path(r"C:\Program Files\BlueStacks\HD-Adb.exe"),
]


class AdbError(RuntimeError):
    pass


class AdbClient:
    """BlueStacks / Android ADB helper. Inputs are ghost (no Windows cursor)."""

    def __init__(self, adb_path: Optional[str] = None, serial: Optional[str] = None) -> None:
        self.adb_path = self._resolve_adb(adb_path)
        self.serial = serial

    @staticmethod
    def _resolve_adb(adb_path: Optional[str]) -> str:
        if adb_path:
            p = Path(adb_path)
            if not p.is_file():
                raise AdbError(f"adb not found: {adb_path}")
            return str(p)
        for cand in DEFAULT_ADB_CANDIDATES:
            if cand.is_file():
                return str(cand)
        which = shutil.which("adb")
        if which:
            return which
        raise AdbError("No BlueStacks HD-Adb.exe / adb found")

    def _base_cmd(self) -> List[str]:
        cmd = [self.adb_path]
        if self.serial:
            cmd += ["-s", self.serial]
        return cmd

    def run(self, *args: str, timeout: float = 30.0, check: bool = True) -> subprocess.CompletedProcess:
        cmd = self._base_cmd() + list(args)
        try:
            proc = subprocess.run(
                cmd,
                capture_output=True,
                timeout=timeout,
                check=False,
            )
        except subprocess.TimeoutExpired as exc:
            raise AdbError(f"adb timeout: {' '.join(cmd)}") from exc
        if check and proc.returncode != 0:
            err = (proc.stderr or b"").decode("utf-8", errors="replace").strip()
            raise AdbError(f"adb failed ({proc.returncode}): {err or ' '.join(cmd)}")
        return proc

    def devices(self) -> List[str]:
        proc = self.run("devices", check=True)
        lines = (proc.stdout or b"").decode("utf-8", errors="replace").strip().splitlines()
        out: List[str] = []
        for line in lines[1:]:
            parts = line.split()
            if len(parts) >= 2 and parts[1] == "device":
                out.append(parts[0])
        return out

    def ensure_device(self) -> str:
        if self.serial:
            return self.serial
        devs = self.devices()
        if not devs:
            raise AdbError("No ADB device online (start BlueStacks)")
        if len(devs) == 1:
            self.serial = devs[0]
            return self.serial
        # Prefer classic emulator serial
        for d in devs:
            if d.startswith("emulator-"):
                self.serial = d
                return self.serial
        self.serial = devs[0]
        return self.serial

    def screencap_png(self) -> bytes:
        self.ensure_device()
        proc = self.run("exec-out", "screencap", "-p", timeout=60.0)
        data = proc.stdout or b""
        if len(data) < 100:
            raise AdbError("screencap returned empty data")
        if data[:8] != b"\x89PNG\r\n\x1a\n":
            # Some adb clients expand LF→CRLF inside the binary stream
            fixed = data.replace(b"\r\n", b"\n")
            if fixed[:8] == b"\x89PNG\r\n\x1a\n":
                return fixed
            raise AdbError("screencap did not return a PNG")
        return data

    def tap(self, x: int, y: int) -> None:
        self.ensure_device()
        self.run("shell", "input", "tap", str(int(x)), str(int(y)))

    def swipe(self, x1: int, y1: int, x2: int, y2: int, duration_ms: int = 300) -> None:
        self.ensure_device()
        self.run(
            "shell",
            "input",
            "swipe",
            str(int(x1)),
            str(int(y1)),
            str(int(x2)),
            str(int(y2)),
            str(int(duration_ms)),
        )

    def wm_size(self) -> tuple[int, int]:
        self.ensure_device()
        proc = self.run("shell", "wm", "size")
        text = (proc.stdout or b"").decode("utf-8", errors="replace")
        # Physical size: 1920x1080
        for token in text.replace("\r", "\n").split():
            if "x" in token and token[0].isdigit():
                w, h = token.split("x", 1)
                return int(w), int(h)
        raise AdbError(f"cannot parse wm size: {text}")
